### ISSUE TITLE
REST API call to SharePoint files only works in root site collection?

### ISSUE THREAD

**Issue by bobbyware** [AzureAD/azure-activedirectory-library-for-js#155]
Scenario: SPA with Web API for Discovery Service. Everything works fine, I can authenticate, fire search queries (SharePoint REST API) and download files in the root site collection.
But, when I download a file in a non-root site collection I get an error: "Invalid audience Uri '<guid>'
, when <guid> is the guid of my Azure Application.
Permissions are set correctly (read in all site collections).

The two requests at that time:

1) OPTIONS:

OPTIONS https://tenant.sharepoint.com/sites/contoso//_api/web/getfilebyserverrelativeurl('/sites/contoso/Documents/2Q2014%20Report%20of%20Expenditures.docx')/$value HTTP/1.1
Accept: */*
....
HTTP/1.1 200 OK
Cache-Control: private, max-age=0
...

2) Get request:

GET https://tenant.sharepoint.com/sites/contoso//_api/web/getfilebyserverrelativeurl('/sites/contoso/Documents/2Q2014%20Report%20of%20Expenditures.docx')/$value HTTP/1.1
Authorization: Bearer 
<<<token>>
Accept: application/json, text/plain, */*
....

HTTP/1.1 401 Unauthorized
Server: Microsoft-IIS/8.5
x-ms-diagnostics: 3000003;reason="Invalid audience Uri '522d7e75-a030-4d10-986c-f6710b8f9df2'.";category="invalid_client"
...

Is this a bug?

**Comment by bobbyware** [comment]
... When I do this, it works!

var authContext = new AuthenticationContext(adalAuthenticationService.config);
authContext.acquireToken($scope.rootSiteServiceEndpointUri, function (error, token) {
...
}

Probably I have to set a config property on the Angular ADAL JS service? Anybody?

**Comment by weijjia** [comment]
Did you check about this sample https://github.com/AzureADSamples/SinglePageApp-WebAPI-AngularJS-DotNet?

**Comment by bobbyware** [comment]
What should I check in that example? I don't have a custom Web API.

**Comment by bobbyware** [comment]
Fixed it myself with a little workaround:

https://rolandoldengarm.wordpress.com/2015/07/27/angular-adal-js-sharepoint-rest-api-in-non-root-site-collections/

**Comment by tushargupta51** [comment]
@bobbyware what does a non-root site url looks like? Are those urls for app's backend? Adal, by default, only fetches tokens for its own backend. For accessing external web apis, endpoints property should be specified at the time, like you explained in your blog entry.

**Comment by tushargupta51** [comment]
If the non-site url has different host than the app's host, then it needs to be specified as one of the endpoints in the adal configuration. I am closing this issue. Feel free to reopen or comment if you have any questions.

**Comment by balbany** [comment]
Could we get a clear code sample on how this can be implemented? Everything I've tried has failed. Updated link to Roland's post on his workaround (original is broken) is here: [http://rolandoldengarm.com/index.php/2015/07/27/angular-adal-js-sharepoint-rest-api-in-non-root-site-collections/](url) but I'm not convinced that this still works with the latest release of adal.js...

Thanks in advance for any help!
